packages/typespec-client-generator-core/src/lib.ts

```diff
@@ -49,6 +49,12 @@ export const $lib = createTypeSpecLibrary({
         default: "Cannot have a union containing only null types.",
       },
     },
+    "union-circular": {
+      severity: "error",
+      messages: {
+        default: "Cannot have a union containing self.",
+      },
+    },
```

Comment:
I think we should open an issue in microsoft/typespec because typespec as a language should never allow this - a type defined in this way is impossible to exist.

Reply:
issue here: https://github.com/microsoft/typespec/issues/6106